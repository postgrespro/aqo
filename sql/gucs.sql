-- Preliminaries
CREATE EXTENSION IF NOT EXISTS aqo;
SELECT true AS success FROM aqo_reset();

-- Utility tool. Allow to filter system-dependent strings from an explain output.
CREATE OR REPLACE FUNCTION expln(query_string text) RETURNS SETOF text AS $$
BEGIN
    RETURN QUERY
        EXECUTE format('%s', query_string);
    RETURN;
END;
$$ LANGUAGE PLPGSQL;

SET aqo.mode = 'learn';
SET aqo.show_details = true;
SET compute_query_id = 'auto';

CREATE TABLE t(x int);
INSERT INTO t (x) (SELECT * FROM generate_series(1, 100) AS gs);
ANALYZE t;

SELECT true AS success FROM aqo_reset();
-- Check AQO addons to explain (the only stable data)
SELECT regexp_replace(
        str,'Query Identifier: -?\m\d+\M','Query Identifier: N','g') as str FROM expln('
  EXPLAIN (ANALYZE, VERBOSE, COSTS OFF, TIMING OFF, SUMMARY OFF)
    SELECT x FROM t;
') AS str;
SELECT regexp_replace(
        str,'Query Identifier: -?\m\d+\M','Query Identifier: N','g') as str FROM expln('
  EXPLAIN (ANALYZE, VERBOSE, COSTS OFF, TIMING OFF, SUMMARY OFF)
    SELECT x FROM t;
') AS str;
SET aqo.mode = 'disabled';

-- Check existence of the interface functions.
SELECT obj_description('aqo_cardinality_error'::regproc::oid);
SELECT obj_description('aqo_execution_time'::regproc::oid);
SELECT obj_description('aqo_drop_class'::regproc::oid);
SELECT obj_description('aqo_cleanup'::regproc::oid);
SELECT obj_description('aqo_reset'::regproc::oid);

\df aqo_cardinality_error
\df aqo_execution_time
\df aqo_drop_class
\df aqo_cleanup
\df aqo_reset

-- Check stat reset
SELECT count(*) FROM aqo_query_stat;
SELECT true AS success FROM aqo_reset();
SELECT count(*) FROM aqo_query_stat;

DROP EXTENSION aqo;
