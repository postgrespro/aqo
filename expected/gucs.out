-- Preliminaries
CREATE EXTENSION IF NOT EXISTS aqo;
SELECT true AS success FROM aqo_reset();
 success 
---------
 t
(1 row)

-- Utility tool. Allow to filter system-dependent strings from an explain output.
CREATE OR REPLACE FUNCTION expln(query_string text) RETURNS SETOF text AS $$
BEGIN
    RETURN QUERY
        EXECUTE format('%s', query_string);
    RETURN;
END;
$$ LANGUAGE PLPGSQL;
SET aqo.mode = 'learn';
SET aqo.show_details = true;
SET compute_query_id = 'auto';
CREATE TABLE t(x int);
INSERT INTO t (x) (SELECT * FROM generate_series(1, 100) AS gs);
ANALYZE t;
SELECT true AS success FROM aqo_reset();
 success 
---------
 t
(1 row)

-- Check AQO addons to explain (the only stable data)
SELECT regexp_replace(
        str,'Query Identifier: -?\m\d+\M','Query Identifier: N','g') as str FROM expln('
  EXPLAIN (ANALYZE, VERBOSE, COSTS OFF, TIMING OFF, SUMMARY OFF)
    SELECT x FROM t;
') AS str;
                      str                       
------------------------------------------------
 Seq Scan on public.t (actual rows=100 loops=1)
   AQO not used
   Output: x
 Query Identifier: N
 Using aqo: true
 AQO mode: LEARN
 JOINS: 0
(7 rows)

SELECT regexp_replace(
        str,'Query Identifier: -?\m\d+\M','Query Identifier: N','g') as str FROM expln('
  EXPLAIN (ANALYZE, VERBOSE, COSTS OFF, TIMING OFF, SUMMARY OFF)
    SELECT x FROM t;
') AS str;
                      str                       
------------------------------------------------
 Seq Scan on public.t (actual rows=100 loops=1)
   AQO: rows=100, error=0%
   Output: x
 Query Identifier: N
 Using aqo: true
 AQO mode: LEARN
 JOINS: 0
(7 rows)

SET aqo.mode = 'disabled';
-- Check existence of the interface functions.
SELECT obj_description('aqo_cardinality_error'::regproc::oid);
                                                obj_description                                                
---------------------------------------------------------------------------------------------------------------
 Get cardinality error of queries the last time they were executed. Order queries according to an error value.
(1 row)

SELECT obj_description('aqo_execution_time'::regproc::oid);
                                                                                                             obj_description                                                                                                              
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Get execution time of queries. If controlled = true (AQO could advise cardinality estimations), show time of last execution attempt. Another case (AQO not used), return an average value of execution time across all known executions.
(1 row)

SELECT obj_description('aqo_drop_class'::regproc::oid);
                       obj_description                        
--------------------------------------------------------------
 Remove info about an query class from AQO ML knowledge base.
(1 row)

SELECT obj_description('aqo_cleanup'::regproc::oid);
               obj_description                
----------------------------------------------
 Remove unneeded rows from the AQO ML storage
(1 row)

SELECT obj_description('aqo_reset'::regproc::oid);
        obj_description         
--------------------------------
 Reset all data gathered by AQO
(1 row)

\df aqo_cardinality_error
                                                                 List of functions
 Schema |         Name          |                                  Result data type                                   | Argument data types | Type 
--------+-----------------------+-------------------------------------------------------------------------------------+---------------------+------
 public | aqo_cardinality_error | TABLE(num integer, id bigint, fshash bigint, error double precision, nexecs bigint) | controlled boolean  | func
(1 row)

\df aqo_execution_time
                                                                 List of functions
 Schema |        Name        |                                    Result data type                                     | Argument data types | Type 
--------+--------------------+-----------------------------------------------------------------------------------------+---------------------+------
 public | aqo_execution_time | TABLE(num integer, id bigint, fshash bigint, exec_time double precision, nexecs bigint) | controlled boolean  | func
(1 row)

\df aqo_drop_class
                            List of functions
 Schema |      Name      | Result data type | Argument data types | Type 
--------+----------------+------------------+---------------------+------
 public | aqo_drop_class | integer          | queryid bigint      | func
(1 row)

\df aqo_cleanup
                                 List of functions
 Schema |    Name     | Result data type |        Argument data types        | Type 
--------+-------------+------------------+-----------------------------------+------
 public | aqo_cleanup | record           | OUT nfs integer, OUT nfss integer | func
(1 row)

\df aqo_reset
                         List of functions
 Schema |   Name    | Result data type | Argument data types | Type 
--------+-----------+------------------+---------------------+------
 public | aqo_reset | bigint           |                     | func
(1 row)

-- Check stat reset
SELECT count(*) FROM aqo_query_stat;
 count 
-------
     1
(1 row)

SELECT true AS success FROM aqo_reset();
 success 
---------
 t
(1 row)

SELECT count(*) FROM aqo_query_stat;
 count 
-------
     0
(1 row)

DROP EXTENSION aqo;
